pkgs/development/interpreters/octave/build-octave-package.nix

```diff
@@ -0,0 +1,103 @@
+# additional configuration provided, and the namePrefix to use (based on the
+# pname and version of Octave), the octave package, etc.
+
+{ lib
+, stdenv
+, config
+, octave
+, texinfo
+, computeRequiredOctavePackages
+, writeRequiredOctavePackagesHook
+}:
+
+# The inner function contains information required to build the individual
+# libraries.
+{ fullLibName ? "${attrs.pname}-${attrs.version}"
+
+, src
+
+, dontPatch ? false
+, patches ? []
+, patchPhase ? ""
+
+# Build-time dependencies for the package, which were compiled for the system compiling this.
+, nativeBuildInputs ? []
+
+# Build-time dependencies for the package, which may not have been compiled for the system compiling this.
+, buildInputs ? []
+
+# Propagate build dependencies so in case we have A -> B -> C,
+# C can import package A propagated by B
+# Run-time dependencies for the package.
+, propagatedBuildInputs ? []
+
+# Octave packages that are required at runtime for this one.
+# These behave similarly to propagatedBuildInputs, where if
+# package A is needed by B, and C needs B, then C also requires A.
+# The main difference between these and propagatedBuildInputs is
+# during the package's installation into octave, where all
+# requiredOctavePackages are ALSO installed into octave.
+, requiredOctavePackages ? []
+
+, meta ? {}
+
+, passthru ? {}
+
+, ... } @ attrs:
+
+let
+  requiredOctavePackages' = computeRequiredOctavePackages requiredOctavePackages;
+
+  self = stdenv.mkDerivation {
+    packageName = "${fullLibName}";
+    # The name of the octave package ends up being
+    # "octave-version-package-version"
+    name = "${octave.pname}-${octave.version}-${fullLibName}";
+    inherit src;
+
+    OCTAVE_HISTFILE = "/dev/null";
+
+    # This states that any package built with the function that this returns
+    # will be an octave package. This is used for ensuring other octave
+    # packages are installed into octave during the environment building phase.
+    isOctavePackage = true;
+
+    inherit dontPatch patches;
+
+    requiredOctavePackages = requiredOctavePackages';
+
+    nativeBuildInputs = [
+      octave
+      writeRequiredOctavePackagesHook
+    ]
+    ++ nativeBuildInputs;
+
+    buildInputs = buildInputs ++ requiredOctavePackages';
+
+    propagatedBuildInputs = propagatedBuildInputs ++ [ texinfo ];
+
+    buildPhase = ''
+      runHook preBuild
+
+      # This trickery is needed because Octave expects a single directory inside
+      # at the top-most level of the tarball.
+      tar --transform 's,^,${fullLibName}/,' -cz * -f ${fullLibName}.tar.gz
+
+      mkdir -p $out
+      octave-cli --eval "pkg build $out ${fullLibName}.tar.gz"
+
+      runHook postBuild
+    '';
+
+    # We don't install here, because that's handled when we build the environment
+    # together with Octave.
+    dontInstall = true;
+
+    inherit meta;
+  } // { inherit patchPhase; };
+in lib.extendDerivation true passthru self
```

Comment:
I think previously [@FRidh said](https://github.com/NixOS/nixpkgs/pull/108562#discussion_r556626751) that exactly this line - `extendDerivation` was a hack needed for python not needed here. I too think we should get by with just `stdenv.mkDerivation` as always.

Reply:
Ahh... I don't quite understand how it is a hack, but OK. I just removed it. Testing and checking it now.